### ISSUE TITLE
workaround nsfw not supporting symlinks by resolving the realpath of that link

### ISSUE THREAD

**Issue by hexa00** [theia-ide/theia#942]
We should call realpath on a watched URI so that we work around the symlink issue ?

**Comment by akosyakov** [comment]
We do it on the root URI: https://github.com/theia-ide/theia/blob/b4dd8fedd3a7740bfc4a305b1698537741a5c156/packages/filesystem/src/node/nsfw-watcher/nsfw-filesystem-watcher.ts#L99

It is not possible from node js code because it will require an initial scan of the whole tree, what actually chokidar does, and will kill performance. The only way is to fix in native code: https://github.com/Axosoft/nsfw/issues/30#issuecomment-329076451

**Comment by marcdumais-work** [comment]
@akosyakov I stumbled on an interesting case: I had created a symlink to my theia source directory, under /home/, back when we had to set the WS each time the client was started or refreshed. Since then, that path has been my workspace root.

When a watched file is modified, under that WS, the F-E is notified that the file has changed, but the  reported changed URI has the symlink path resolved, and so the FE doesn't realize it's the same file it's watching, when comparing them. 

What's the best way to deal with this? Should we expose "realpath()" in the filesystem API, so that a frontend extension can also resolve such a path? This way the FE could do that first, and so know in advance, what the "real", resolved URI, is.

**Comment by hexa00** [comment]
Or maybe a better option the this.fireDidFilesChanged(), event could carry the watcherId and we could check for that instead of checking for the files ?

**Comment by hexa00** [comment]
Even better would be to have the filtering done inside the nsfw-filesystem-watcher so that when you register events on a watcher you only get the events for that watcher

**Comment by akosyakov** [comment]
What if the watcher rebases changed URIs to the original base?

**Comment by marcdumais-work** [comment]
So the client would be notified with the original URI it use to register the watch? That would work too.

**Comment by hexa00** [comment]
But in general maybe there's something I'm missing but if I do something like:

watcher1 = watch('/test')

And test has the following structure:

```
test/a
test/b
test/c
```

And a do another watch with:

watcher2 = watch('foo')

```
foo/c
foo/d
foo/e
```

Every watcher will get notified of file changes so that means the code of watcher1 has to list recursively all the files before adding the watch to be able to look them up once it receives a notification ?

Like a change in foo/c will trigger an event on watcher1  ? and watcher1 has to check if it matches 'test/*' ?

That seems like it's not a good idea.... ?

**Comment by akosyakov** [comment]
@hexa00 sorry, I could not grasp the question? How `test` and `foo` related?

**Comment by hexa00** [comment]
@akosyakov they are both watched...

I mean it seems that the notifications are not watcher specific... so the watcher client needs to check if the notification are addressed to itself.

But this should be done upstream so that a watcher only receives events about what it asked to be watched ? And not about what the other watchers are watching...

**Comment by akosyakov** [comment]
@hexa00 ok, I think, I understood it, I will add a callback to `watch` method

How is it related to this issue?

**Comment by hexa00** [comment]
@akosyakov because @marcdumais-work is checking for the path he asked to watch when receiving the notifications because he's receiving the notifications from all the watchers.

But if he was receiving only the notification he asked for this check would not be needed so there would be no problem about the cannonical path vs the path with the symlink.

**Comment by hexa00** [comment]
The issue is about https://github.com/theia-ide/theia/issues/942#issuecomment-349081477 more than what I originally reported

**Comment by akosyakov** [comment]
@hexa00 ok, could you point me to the code? It helps me to understand better.

**Comment by akosyakov** [comment]
I don't think it is going to help if a client pass a symlink to watch, since changed URIs could be resolved by nfsw and different base. I think rebasing to the original base is necessary.

**Comment by hexa00** [comment]
I don't mind if we resolve it back to the original base.

But I don't understand why when you add a watcher it reports the notifications of all the watchers... ?

**Comment by hexa00** [comment]
And since he's checking only for 1 file we would not need any check at all and not care about the base...

**Comment by akosyakov** [comment]
> But I don't understand why when you add a watcher it reports the notifications of all the watchers... ?

Originally it was that one client calls `FileSystemWatcher.watch` to start a workspace watching and other clients subscribe to it with `FileSystemWatcher.onFilesChanged`. It does now work anymore nicely since there are clients which want to watch only particular file or directory. I will add an optional callback to `FileSystemWatcher.watch` for such clients so only they will receive changes and they won't receive changes from other watchers.

**Comment by hexa00** [comment]
Humm why not specifiy the id returned by watch in FileSystemWatcher.onFilesChanged ?

**Comment by hexa00** [comment]
and make the id optional so that if not specified you get it all.. but if it is you get only the events for your watcher..

**Comment by akosyakov** [comment]
I don't want clients force do filtering and want to stick to disposable pattern, client will need do only one thing:
```ts
toDispose.push(await watcher.watch(uri, changes => this.handleFileChanges(changes))
```

**Comment by hexa00** [comment]
I think that's weird since you will have 2 apis that seems like they mean the same thing..

Why use the callback or the onFilesChanged() event, you have to convey that onFileChanged is not filtered while the callback is through documentation...

It seems to me it would be simpler to only have 1 interface and the parameters of that interface specify what you want to be notified on

so if you need filtering you would do :

```typescript
const watcherId = await watcher.watch(uri)
watcher.onFileChanged(changes => this.handleFileChanges(changes), watcherId) ;
toDispose.push(watcherId);
```

And if you don't care about it you would simply do:

```typescript
toDispose.push(await watcher.watch(uri))
watcher.onFileChanged(changes => this.handleFileChanges(changes) ) ;
``` 

Seems more consistant  to me than to do:

```typescript
toDispose.push(await watcher.watch(uri, changes => this.handleFileChanges(changes))
```

If I care about filtering and:

```typescript
toDispose.push(await watcher.watch(uri);
watcher.OnFileChange(changes => this.handleFileChanges(changes))
```

If I don't

**Comment by akosyakov** [comment]
There are multiple clients who want watch workspace changes and they are not responsible for starting watching so they don't have `watcherId`, some of these clients are not even aware of workspace.

**Comment by akosyakov** [comment]
`watcherId` was a number to match watcher objects in the frontend and backend, we could have the following API:
```ts
watcher.watch(uri): Watcher;

interface Watcher implements Disposable {
   onChanged: Event<FileChange>
}
```

but it does not solve the issue from the previous comment as well.

**Comment by akosyakov** [comment]
Reread your proposals, so actually in your case, such clients won't pass `watcherId` to watch changes from all clients, which is also bad because they are not interested in all.

There are 3 cases:
(1) one client watches exclusive something and does not affect others
```ts
const watcher = await watcher.watch(uri, changes => this.handleFileChanges(changes));
watcher.dispose();
```
vs
```ts
// bogus affects others
const watcher = await watcher.watch(uri);
watcher.onChanged(changes => this.handleFileChanges(changes), watcher);
// have to implement and preserve mapping between number and disposable internally
// disposable pattern could not be used as API
watcher.unwatch(watcher);
```
(2) multiple clients watching fs changes
```ts
watcher.handleFileChanges(changes => this.handleFileChanges(changes));
```
vs
```ts
// bogus because it will get changes from (1) not only from (3)
watcher.onChanged(changes => this.handleFileChanges(changes));
```
(3) one client starts fs watching
```ts
const watcher = await watcher.watch(uri);
watcher.dispose();
```
vs
```ts
const watcher = await watcher.watch(uri);
watcher.unwatch(watcher);
```

**Comment by hexa00** [comment]
> which is also bad because they are not interested in all.

I don't understand ? if no watcherId is passed they are interested about all changes.

I think you're api is what I'm looking for:

```typescript
watcher.watch(uri): Watcher;

interface Watcher implements Disposable {
   onChanged: Event<FileChange>
}
```
But if you call watcher.onChanged() directly then you get all events. So it's clear that:

aWatcher.onChanged(): is specific to the watch you just made
singletonWatcher.onChanged: is all the files watched by the watcher singleton

**Comment by akosyakov** [comment]
ok, agreed, it will require quite a big refactoring to decouple workspace <-> filesystem packages.
I will do the simple thing now with callback and later properly refactor it with multiple watchers.

**Comment by hexa00** [comment]
Ok thx

**Comment by sgraband** [comment]
nsfw replaced by `@parcel/watcher` which natively handles symlinks via realpath resolution.
